src/Input/Tokenizer.php

```diff
@@ -0,0 +1,464 @@
+            );
+            if (strlen($arg) > 1) {
+                $arg = ltrim($arg, Token::TOKEN_VARIADIC_O);
+            } else {
+                return $tokens;
+            }
+        }
+
+        if ($this->isConstant($arg)) {
+
+            if ($this->isVariadicSymbol($arg[strlen($arg) - 1] ?? '')) {
+
+                $tokens[] = new Token(Token::TYPE_CONSTANT, rtrim($arg, Token::TOKEN_VARIADIC_C));
+                $tokens[] = new Token(
+                    Token::TYPE_VARIADIC, 
+                    Token::TOKEN_VARIADIC_C
+                );
+            } else {
+                $tokens[] = new Token(Token::TYPE_CONSTANT, $arg);
+            }
+            return $tokens;
+        }
+
+        if ($this->isLiteralSymbol($arg)) {
+            $tokens[] = new Token(Token::TYPE_LITERAL, $arg);
+            return $tokens;
+        }
+
+        if ($this->isShortOption($arg)) {
+            $tokens[] = new Token(Token::TYPE_SHORT, $arg);
+            return $tokens;
+        }
+
+        if ($this->isPackedOptions($arg)) {
+            $t = array_map(function($arg) {
+                return new Token(
+                    Token::TYPE_SHORT, 
+                    Token::TOKEN_OPTION_SHORT . $arg
+                );
+            }, str_split(ltrim($arg, Token::TOKEN_OPTION_SHORT)));
+            array_push($tokens, ...$t);
+            return $tokens;
+        }
+
+        if ($this->isShortEqOptions($arg)) {
+            $parts = explode(Token::TOKEN_OPTION_EQ, $arg, 2);
+            $tokens[] = new Token(Token::TYPE_SHORT, $parts[0]);
+            if (!empty($parts[1])) {
+                $t = $this->tokenize($parts[1]);
+                array_push($tokens, ...$t);
+            }
+            return $tokens;
+        }
+
+        if ($this->isLongOption($arg)) {
+            $tokens[] = new Token(Token::TYPE_LONG, $arg);
+            return $tokens;
+        }
+
+        if ($this->isLongEqOption($arg)) {
+            $parts = explode(Token::TOKEN_OPTION_EQ, $arg, 2);
+            $tokens[] = new Token(Token::TYPE_LONG, $parts[0]);
+
+            if (!empty($parts[1])) {
+                $t = $this->tokenize($parts[1]);
+                array_push($tokens, ...$t);
+            }
+
+            return $tokens;
+        }
+
+        // Unclassified, treat as constant:
+        return [new Token(Token::TYPE_CONSTANT, $arg)];
+
+    }
+
+    /**
+     * Get the current token.
+     * For Iterator interface.
+     * 
+     * @return Token
+     */
+    public function current(): Token
+    {
+        return $this->tokens[$this->index];
+    }
+    
+    /**
+     * Get the next token.
+     * Without moving the pointer.
+     * 
+     * @param int $offset
+     * 
+     * @return Token|null
+     */
+    public function offset(int $offset): ?Token
+    {
+        if (isset($this->tokens[$this->index + $offset])) {
+            return $this->tokens[$this->index + $offset];
+        }
+        return null;
```

Comment:
oneliner with null coalesce:

`return $this->tokens[$this->index + $offset] ?? null;`